### ISSUE TITLE
Invalid memory access inside allreduce_intra_recursivedoubling

### ISSUE THREAD

**Issue by patflick** [open-mpi/ompi#1462]
**TLDR:**

The buffer for reductions gets allocated to size: `true_extend + (count-1)*extend`.
(For <= 1.10.* this was hardcoded in the `malloc` call in the reduction functions, in the current code status this is calculated inside the `opal_datatype_span()` function.)
This can lead to faulty memory accesses when a custom reduction operation casts the pointers to its C/C++ type and due to default copy construction/assignment tries to copy the whole datatype to its full extend (rather than the un-padded true-extend).

**Example:**
```c++
#include <utility>
#include <mpi.h>

typedef std::pair<size_t, int> p_t;

void mympi_op(void* invec, void* inoutvec, int* len, MPI_Datatype* dt) {
    p_t* in = static_cast<p_t*>(invec);
    p_t* inout = static_cast<p_t*>(inoutvec);
    for (int i = 0; i < *len; ++i) {
        if (in->first < inout->first) {
            *inout = p_t(*in);
        }
    }
}

int main(int argc, char *argv[])
{
    MPI_Init(&argc, &argv);
    // create a type for the pair
    MPI_Datatype tmp, my_type;
    int blen[2] = {1, 1};
    MPI_Aint displ[2];
    displ[0] = offsetof(p_t, first);
    displ[1] = offsetof(p_t, second);
    MPI_Datatype types[2] = {MPI_UNSIGNED_LONG, MPI_INT};
    MPI_Type_create_struct(2, blen, displ, types, &tmp);
    MPI_Type_create_resized(tmp, 0, sizeof(p_t), &my_type);
    MPI_Type_commit(&my_type);
    // create custom op
    MPI_Op myop;
    MPI_Op_create(&mympi_op, 1, &myop);
    // prepare pairs
    int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    std::pair<size_t, int> p(13*rank, rank);
    std::pair<size_t, int> q;
    // allreduce: produces valgrind errors
    MPI_Allreduce(&p, &q, 1, my_type, myop, MPI_COMM_WORLD);
    MPI_Finalize();
    return 0;
}
```

When compiled with gcc 4.9.2 (without optimization) using the mpic++ wrapper from OpenMPI 1.10.2 and executed within valgrind, this will output the following error (multiple times):
```
==19706== Invalid read of size 8
==19706==    at 0x4085AD: mympi_op(void*, void*, int*, ompi_datatype_t**) (in /home/patrick/dev/c++/tests/a.out)
==19706==    by 0xC85E3A6: ompi_coll_tuned_allreduce_intra_recursivedoubling (in /usr/local/lib/openmpi/mca_coll_tuned.so)
==19706==    by 0x50AB5B2: PMPI_Allreduce (in /usr/local/lib/libmpi.so.12.0.0)
==19706==    by 0x408732: main (in /home/patrick/dev/c++/tests/a.out)
==19706==  Address 0xd669278 is 8 bytes inside a block of size 12 alloc'd
==19706==    at 0x4C2BBA0: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==19706==    by 0xC85E028: ompi_coll_tuned_allreduce_intra_recursivedoubling (in /usr/local/lib/openmpi/mca_coll_tuned.so)
==19706==    by 0x50AB5B2: PMPI_Allreduce (in /usr/local/lib/libmpi.so.12.0.0)
==19706==    by 0x408732: main (in /home/patrick/dev/c++/tests/a.out)
```

The assembly shows that gcc compiles the std::pair copy operation via two 8 bytes `mov` operations.
The `sizeof` of the pair type is 16 bytes due to the 8 byte alignment, however, the custom datatype's true-extend and extend are 12 and 16 respectively.
For all current versions of OpenMPI, the buffer allocated is of size 12. When copied within a custom reduction operation, gcc might align copies to 8 bytes, thus resulting in a 4 byte invalid read.

If executed on larger vectors/arrays of these pairs, the program above can segfault.

**Final notes**

Note that the example code above is a minimal example to reproduce the problem. We found the issue in a much larger project, which segfaulted within the `allreduce_intra_recursivedoubling()` function for much larger reductions.

Is there a reason of why the `span` should not be allocated to the full `count*extend`?

**Comment by ggouaillardet** [comment]
@patflick the allocated size ```true_extend + (count-1)*extend``` is correct and you can find a lengthy explanation in ```ompi/mca/coll/basic/coll_basic_reduce.c```
if you simply use ```count*extend```, then you do not support the case when ```extend``` < ```true_extend```.

that being said, using ```max(count*extend, true_extend + (count-1)*extend)``` will likely fix your issue.

honestly, i do not whether this should be handled in OpenMPI and how ...
@bosilca any thoughts ?

for the time being, i can only suggest you redefine your datatype to make it aware the padding is used by the compiler with something like


```c
    MPI_Datatype my_type;
    p_t foo;
    int blen[3] = {1, 1, 0};
    MPI_Aint displ[3];
    displ[0] = offsetof(p_t, first);
    displ[1] = offsetof(p_t, second);
    displ[2] = offsetof(p_t, second) + sizeof(foo.second);
    blen[2] = sizeof(foo) - displ[2];
    MPI_Datatype types[3] = {MPI_UNSIGNED_LONG, MPI_INT, MPI_BYTE};
    MPI_Type_create_struct(3, blen, displ, types, &my_type);
    MPI_Type_commit(&my_type);
```

**Comment by patflick** [comment]
@ggouaillardet that's a good point, i didn't think about the `extent < true_extent` case.

There is no explanation in `coll_basic_reduce.c`, are you referring to the description at https://github.com/open-mpi/ompi/blob/master/opal/datatype/opal_datatype.h#L332 ?

I'm a bit reluctant to adding explicit padding to the MPI datatype, since that would destroy the whole point of using MPI datatypes (rather than contiguous `MPI_BYTE`s) in the first place (i.e., support on heterogeneous systems). My code is being used as a MPI wrapper in other projects, thus this property would propagate to those projects as well.

I'm thinking of creating my own buffers instead and manually copying the `true_extent` of the datatype (as returned by `MPI_Type_get_true_extent`).
If there is more than one item, I'm assuming the last item in the given `invec` and `inoutvec` buffers is the one that has its padding "cut off"? Is that valid?

**Comment by ggouaillardet** [comment]
@patflick this is in v1.10 at https://github.com/open-mpi/ompi-release/blob/v1.10/ompi/mca/coll/basic/coll_basic_reduce.c
this section was moved to an other place in master

i do not understand what would be the issue with heterogenous systems here
(e.g. one long, one int and four bytes). one long and two int might be faster to handle internally though.

from a performance point of view, and on a homogenous cluster, OpenMPI considers the type with explicit added padding as a contiguous one, and that is much easier and faster to move than the initial one, that is non contiguous.

out of curiosity, what happens if you manually set both fields of a ```p_t``` and copy the object into an other one ? will valgrind complain because it copies 16 bytes that do include 4 unitialized bytes ?
does this depend on the level of optimization ?

from an OpenMPI point of view, should we roundup the allocated memory size just in case the compiler optimization does use some extra padding ?
i am waiting for a second opinion before i answer that

**Comment by patflick** [comment]
The explanation makes sense, thanks for pointing to where to find it.

> i do not understand what would be the issue with heterogeneous systems here
(e.g. one long, one int and four bytes). one long and two int might be faster to handle internally though.

Some systems may not pad the pair by 4 bytes. Thus making the datatypes created on two systems (one that pads, and one that doesn't) incompatible? I'm not a 100% sure on this :)

I'll test some different ways of copying/assignment and different optimization levels and let you know which ones work. Manually assigning the fields of the pair or a struct is not really an option for me, since I'm implementing a general and C++ templated solution that can't assume anything on the given type.

**Comment by ggouaillardet** [comment]
@patflick ah, i see
if the size of a type is 16 on one system (because of padding) but 12 on an other (because of no padding), then yes, this is an issue.
as far as i am concerned, the easiest way is to manually add the padding in the C type so your type is contiguous and its first long is always aligned.

**Comment by patflick** [comment]
Unfortunately, that is not possible for me, since my code is performing reductions on arbitrary C++ types (templated types). I can thus assume nothing about the data layout, padding, or size of the datatype.

**Comment by patflick** [comment]
@ggouaillardet thanks for helping me understand the problem. I now implemented the following solution in my code. This now works, since it explicitly limits the reading/writing to the true_extend of the datatype.

```c++
    template <typename Func>
    static void custom_function(Func func, void* invec, void* inoutvec, int* len, MPI_Datatype* dt) {
        if (*len > 1) {
            T* in = (T*) invec;
            T* inout = (T*) inoutvec;
            for (int i = 0; i < *len-1; ++i) {
                inout[i] = func(in[i], inout[i]);
            }
        }
        // only read and write the `true_extent` of the datatype
        // for the last item (otherwise we might encounter a memory error)
        // [see github OpenMPI issue #1462]
        T in_buf;
        T inout_buf;
        MPI_Aint true_extent, true_lb;
        MPI_Type_get_true_extent(*dt, &true_lb, &true_extent);
        memcpy((char*)&in_buf+true_lb, (char*)invec+(sizeof(T)*(*len-1))+true_lb, true_extent);
        memcpy((char*)&inout_buf+true_lb, (char*)inoutvec+(sizeof(T)*(*len-1))+true_lb, true_extent);
        // now do the operation on our local buffers
        inout_buf = func(in_buf, inout_buf);
        // copy the results back, again only to true_extent
        memcpy((char*)inoutvec+(sizeof(T)*(*len - 1))+true_lb, (char*)&inout_buf+true_lb, true_extent);
    }
```
(from https://github.com/patflick/mxx/blob/master/include/mxx/reduction.hpp#L264 )

**Comment by patflick** [comment]
@ggouaillardet 

> out of curiosity, what happens if you manually set both fields of a p_t and copy the object into an other one ? will valgrind complain because it copies 16 bytes that do include 4 unitialized bytes ?
does this depend on the level of optimization ?

I did some further tests on multiple different variations of the copy/assignment and optimization levels. Interestingly enough, only one certain combination results in valgrind complaining. I don't really know how g++ handles the different cases internally, so I can't really explain these result quite yet.

Variant1:
```c++
*inout = p_t(*in);
```
Memory errors only with `-O0`. Turning on optimization (`-O1-3`) gets rid of the problem.

Variant 2
```
*inout = *in;
```
No problems for any `-O0-3`

Variant 3:
```c++
inout->first = in->first;
inout->second = in->second;
```
No problems for any `-O0-3`

**Comment by bosilca** [comment]
@patflick the assumption of the MPI_Op is that you do not touch the memory outside the layout described by the tuple (MPI_Datatype, count). Thus, doing a full memcpy with total disregard for the memory layout is incorrect.

**Comment by patflick** [comment]
@bosilca So without assuming anything about the memory layout of the datatype, could I use/abuse a self-send (MPI_Sendrecv to MPI_COMM_SELF) to safely copy the data from the MPI buffer into my own prior to accessing it?

**Comment by ggouaillardet** [comment]
@patflick if you want to copy a buffer into a larger one, then yes, ```MPI_Sendrecv``` can be used.

@bosilca could we simply align the size of the allocated memory ?
there is already a wrapper to align memory, it could be easy to extend it to fix the size
(e.g. malloc 16 bytes when 12 are requested)
that will not make valgrind happy, but that would avoid a crash due to compiler optimization

**Comment by patflick** [comment]
@bosilca Due to how the custom operators are used in my application, I can assume that each item (of the `count` total items) is of a fixed size type `T` (such as a basic type or `struct`/`class`). I can exclude spaced/strided datatypes due to how the code is used. Thus I can assume at least that: `sizeof(T) == extent(data_type) >= true_extent(data_type)`.

I understand that my current solution as given above is specialized to the current behavior in OpenMPI.  I I guess I shouldn't assume that the datatype is not further packed into the buffer.

I can think of two possible solutions:
1) Creating my own buffers of size `len*sizeof(T)` and using `MPI_Sendrecv` as the "copy" operation.

This has the downside of requiring me to copy both input vectors to it's full length, and after the combination operation to copy the result back (using `MPI_Sendrecv` as the copy),
rather than just the last element as before.

2) For a very general solution, would I could decode the datatype with `MPI_Type_get_envelope` and it's related functions. Is this a viable option?

Do you have any better ideas? My current "solution" (only treat the last element in the buffer specially) seems to work with all MPI implementations I have tested it on so far (incl. current versions of OpenMPI, MPICH, MVAPICH).

**Comment by bosilca** [comment]
@patflick I would not call an MPI communication in an MPI_Op, even if the communication is with self. While it might work in most cases in Open MPI, I would not guarantee that it is the same in all MPI implementations. Calling MPI_Pack/MPI_Unpack might be safe(r). There is nothing in the MPI standard that would specifically allow for such behavior (also there is nothing that would prevent it).

@ggouaillardet I am not sure what we want to align ? We could trick the allocation to make sure it's fix byte of the data has a similar alignment to the original data. Another solution would be to never try to be smart when we are using a user provided MPI_Op, but this will have the negative side-effect of forcing us to modify all the collective algorithms.

**Comment by bosilca** [comment]
I understand your problem, you don't want to care about the memory layout described by the datatype, but instead you want to simply copy the entire extent. If you do not care about possible valgrind complaints about using non-initialized memory (basically all the gaps in any temporary array will be non-initialized), then the solution you proposed above works (I'm not saying it is correct, but will work in all cases). The reason why this solution works in all MPI implementations is that based on assumption that the MPI_Op does not modify outside the memory layout described by the datatype, the only possible optimization an MPI implementation can do is to minimize the size of the allocated memory by removing the first and the last gap (no extra packing is possible).

If would think that the solution based on ```MPI_Type_get_envelope``` is more MPI compliant, but much more costly and time consuming to implement.

**Comment by patflick** [comment]
@bosilca 

Just to understand your suggestion with MPI_Pack/MPI_Unpack, are you suggesting something like the following?

```c++
void custom_function(Func func, void* invec, void* inoutvec, int* len, MPI_Datatype* dt) {
    // Copy `invec` into our own buffer:
    int pack_size;
    MPI_Pack_size(*len, *dt, MPI_COMM_SELF, &pack_size);
    T* my_invec = malloc(*len*sizeof(T));
    char* my_packed_buf = malloc(pack_size);
    int pos = 0;
    MPI_Pack(invec, *len, *dt, my_packed_buf, pack_size, &pos, MPI_COMM_SELF);
    pos = 0;
    MPI_Unpack(my_packed_buf, pack_size, &pos, my_invec, *len, *dt, MPI_COMM_SELF);

   // do the same of inoutvec
   // perform reduction
   // copy back with MPI_Pack + MPI_Unpack
}
```

This will require quite a large additional memory requirement, so I wouldn't be too happy about this solution.

**Comment by bosilca** [comment]
I agree with you. It is a totally suboptimal, and undesirable, solution, but a correct one ;)

**Comment by patflick** [comment]
Thank you @bosilca for your help and suggestions. It's very appreciated!

I'll look into the performance penalty when using Pack/Unpack for
- all elements at once
- one element at a time (so I don't need to allocated additional memory beyond the size of one element)
- maybe in a blocked fashion (tradeoff between mem requirement and computing)

But for now (at least until the next deadline is over), I'll stick with the previously mentioned "work-around" (memcpy `true_extent`).
